odoo/fields.py

```diff
@@ -3624,6 +3625,18 @@ def prefetch_x2many_ids(record, field):
     return unique(id_ for ids in ids_list for id_ in ids)
 
 
+def apply_required(cr, model_name, field_name):
+    """ Set a NOT NULL constraint on the given field, if necessary. """
+    # retrieve the field
+    registry = odoo.registry(cr.dbname)
+    Model = registry[model_name]
+    field = Model._fields[field_name]
+
+    # set the NOT NULL constraint if it still applies
+    if field.store and field.required:
+        sql.set_not_null(cr, Model._table, field_name)
```

Comment:
If `apply_required` is called by `registry.finalize_constraints()`, the registry and the model's class are still the same, but the `_fields` of the model may have been regenerated.  Therefore, I think the function can be simplified to
```suggestion
def apply_required(model, field_name):
    """ Set a NOT NULL constraint on the given field, if necessary. """
    # At the time this function is called, the model's _fields may have been
    # reset, although the model's class is still the same.  Retrieve the field
    # to see whether the NOT NULL constraint still applies.
    field = model._fields[field_name]
    if field.store and field.required:
        sql.set_not_null(model.env.cr, model._table, field_name)
```
Can you adapt the rest of the code, and check that it works?
Thanks!

Reply:
@rco-odoo Just tested it, it works :+1: